### PR TITLE
Refactor fields (step 3) 

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

```SUMMARY: Infrastructure "Refactor fields (step 3)"```

#### Purpose of change

Unhardcode data structures to allow field type customization/addition via json.

#### Describe the solution

Current pull request does the following:

- utilizes new `field_type` class all over the code (and drops `field_t` class);
- minor changes to `field_type` class and json;
- minor changes to internal map editor related to fields;
- rename of some field related functions and variables (i.e. used `density` instead of `intensity` everywhere for consistency).

#### Additional context

Current pull request also contains changes from Step 1 (#31310) and Step 2 (#31311).

Further refactoring will include splitting of dreadful code from `map_field` to smaller functions.

### CODE REVIEW

**Comment by kevingranade** [comment]
build error

**Comment by ZhilkinSerg** [comment]
> build error

That was stupid manual rename. Anyway I plan to fix outlined issues and refactor small things in current PR a bit, so let it hang for a while.

File: src/vehicle.h [hunk 1]
```diff
@@ -1109,8 +1110,8 @@ class vehicle
         // Get maximum velocity for the current movement mode
         int safe_velocity( bool fueled = true ) const;
 
-        // Generate smoke from a part, either at front or back of vehicle depending on velocity.
-        void spew_smoke( double joules, int part, int density = 1 );
+        // Generate field from a part, either at front or back of vehicle depending on velocity.
+        void spew_field( double joules, int part, field_type_id type = fd_null, int density = 1 );
```
Comment (BevapDin) [inline src/vehicle.h]: The `type` is optional and defaults to `fd_null`. When would I leave out the type? Does it even make sense to spawn a `fd_null` field? Is that even possible?
```suggestion
        void spew_field( double joules, int part, field_type_id type, int density = 1 );
```

File: src/map.h [hunk 2]
```diff
@@ -1415,6 +1414,9 @@ class map
         void rad_scorch( const tripoint &p, const time_duration &time_since_last_actualize );
         void decay_cosmetic_fields( const tripoint &p, const time_duration &time_since_last_actualize );
 
+        void process_fire_in_submap( const tripoint &p, const maptile &map_tile, int &part,
```
Comment (BevapDin) [inline src/map.h]: This function does not seem to be implemented.

Comment (ZhilkinSerg) [inline src/map.h]:
```suggestion
```

File: src/map.h [hunk 3]
```diff
@@ -1415,6 +1414,9 @@ class map
...
+                                     field_entry &cur, bool &dirty_transparency_cache, field_entry *tmpfld );
```
Comment (ZhilkinSerg) [inline src/map.h]:
```suggestion
```

File: src/map.h [hunk 4]
```diff
@@ -1415,6 +1414,9 @@ class map
...
+
```
Comment (ZhilkinSerg) [inline src/map.h]:
```suggestion
```

File: src/map.h (hunk 2 above)
Comment (ZhilkinSerg) [reply src/map.h]: Thanks - that were leftovers of refactor attempt.

**Comment by ZhilkinSerg** [comment]
This should be ready if CI succeeds, but I can split rename part from here to a later pull request if it would make reviewing easier.

File: src/material.cpp [hunk 5]
```diff
@@ -93,22 +93,21 @@ void material_type::load( JsonObject &jsobj, const std::string & )
...
+        // If not specified, supply default
+        if( _fire_resist <= 0 ) {
             mat_burn_data mbd;
-            if( flammable ) {
-                mbd.burn = 1;
-            }
-
-            _burn_data[ intensity ] = mbd;
+            mbd.burn = 1;
+            _burn_data.emplace_back( mbd );
```
Comment (alanbrady) [inline src/material.cpp]: This should fix the segfault

```suggestion
            mat_burn_data mbd;
            if( _fire_resist <= 0 ) {
                    mbd.burn = 1;
            }
            _burn_data.emplace_back( mbd );
```

File: src/lightmap.cpp [hunk 6]
```diff
@@ -124,44 +125,33 @@ bool map::build_transparency_cache( const int zlev )
...
-                                break;
+                            } else if( density == 2 ) {
+                                value *= 10;
+                            }
+                        } else if( type == fd_nuke_gas ) {
+                            value *= 10;
+                        } else if( type == fd_fire ) {
+                            value *= 1.0 - ( density * 0.3 );
+                        } else {
+                            value = LIGHT_TRANSPARENCY_SOLID;
```
Comment (alanbrady) [inline src/lightmap.cpp]: Was there something wrong with the switch case?  switch is way easier to grok than a bunch of if's

Comment (alanbrady) [reply src/lightmap.cpp]: Not to mention `switch` is generally more performant than `if/else`

File: src/iuse.cpp [hunk 7]
```diff
@@ -4167,21 +4168,13 @@ int iuse::gasmask( player *p, item *it, bool t, const tripoint &pos )
...
-                        break;
-                    default:
-                        break;
+                const field_type_id fid = entry.get_type();
+                if( fid == fd_smoke ) {
+                    it->set_var( "gas_absorbed", it->get_var( "gas_absorbed", 0 ) + 12 );
+                }
+                if( fid == fd_tear_gas || fid == fd_toxic_gas || fid == fd_gas_vent ||
+                    fid == fd_smoke_vent || fid == fd_relax_gas || fid == fd_fungal_haze ) {
+                    it->set_var( "gas_absorbed", it->get_var( "gas_absorbed", 0 ) + 15 );
```
Comment (alanbrady) [inline src/iuse.cpp]: Ditto here, why scramble the `switch`?

File: src/lightmap.cpp (hunk 6 above)
Comment (ZhilkinSerg) [reply src/lightmap.cpp]: Unfortunately switch only works with basic types, which `int_id<T>` is not. I have some thoughts on fixing this monster function later anyway.

File: src/iuse.cpp (hunk 7 above)
Comment (ZhilkinSerg) [reply src/iuse.cpp]: I will refactor this later. The approach would be to have some nodes in field type definitions which determine processing of fields, i.e.:

Field types `fd_tear_gas`, `fd_toxic_gas`, `fd_gas_vent`, `fd_smoke_vent`, `fd_relax_gas` and `fd_fungal_haze` (and any new similar fields) will have some node in json (e.g. `"field_processing_rule_values": { { "item_gas_absorbed_delta": 15 } }`) which will indicate they would need to go with ```it->set_var( "gas_absorbed", it->get_var( "gas_absorbed", 0 ) + ft.field_processing_rule_values.item_gas_absorbed_delta );```.

File: tests/map_helpers.cpp [hunk 8]
```diff
@@ -60,11 +60,11 @@ void clear_fields( const int zlevel )
     for( int x = 0; x < mapsize; ++x ) {
         for( int y = 0; y < mapsize; ++y ) {
             const tripoint p( x, y, zlevel );
-            std::vector<field_id> fields;
+            std::vector<field_type_id> fields;
```
Comment (ZhilkinSerg) [inline tests/map_helpers.cpp]:
```suggestion
            std::vector<field_id> fields;
```

File: tests/map_helpers.cpp [hunk 9]
```diff
@@ -60,11 +60,11 @@ void clear_fields( const int zlevel )
...
             for( auto &pr : g->m.field_at( p ) ) {
                 fields.push_back( pr.second.get_field_type() );
             }
-            for( field_id f : fields ) {
+            for( field_type_id f : fields ) {
```
Comment (ZhilkinSerg) [inline tests/map_helpers.cpp]:
```suggestion
            for( field_id f : fields ) {
```

**Comment by ZhilkinSerg** [comment]
Current state it probably the smallest I can get (maybe a few more renames are feasable, but it would be just like 15 or 20 changed LOC less).

Monster field processing function can and would be split later, but I really cannot start doing this before field_id enums are changed to int_id<field_type>.

**Comment by kevingranade** [comment]
You grew some conflicts with the vehicle terrain manipulation PR.

**Comment by ZhilkinSerg** [comment]
> 
> 
> You grew some conflicts with the vehicle terrain manipulation PR.

Should be good now. Also I think non-whitespace changes are manageable for review now:

image

File: src/emit.h [hunk 10]
```diff
@@ -60,8 +63,8 @@ class emit
 
     private:
         emit_id id_;
-        field_id field_ = fd_null;
-        int intensity_ = 3;
+        field_type_id field_ = fd_null;
+        int intensity_ = 1;
```
Comment (ZhilkinSerg) [inline src/emit.h]: Apparently this requires changes to `emissions.json`, so existing emissions without `intensity` defined should have `intensity` set to `3`

**Pull request merged**